sdk/core/azure-core/src/uuid.cpp

```diff
@@ -83,19 +96,77 @@ namespace Azure { namespace Core {
     // values are reserved for backward compatibility. The C|c, D|d values are reserved for
     // Microsoft, and the E|e, F|f values are reserved for future use.
     // Therefore, we have to zero out the two high bits, and then set the highest bit to 1.
-    uuid[8] = (uuid[8] & 0x3F) | 0x80;
+    result.m_uuid.data()[8] = (result.m_uuid.data()[8] & 0x3F) | 0x80;
 
-    constexpr uint8_t version = 4; // Version 4: Pseudo-random number
-
-    uuid[6] = (uuid[6] & 0xF) | (version << 4);
+    {
+      constexpr std::uint8_t Version = 4; // Version 4: Pseudo-random number
+      result.m_uuid.data()[6] = (result.m_uuid.data()[6] & 0xF) | (Version << 4);
+    }
 
-    return Uuid(uuid);
+    return result;
   }
 
-  Uuid Uuid::CreateFromArray(std::array<uint8_t, UuidSize> const& uuid)
+  namespace {
+    constexpr size_t UuidStringLength = 36; // 00000000-0000-0000-0000-000000000000
+    constexpr bool IsDashIndex(size_t i) { return i == 8 || i == 13 || i == 18 || i == 23; }
+
+    constexpr std::uint8_t HexToNibble(char c) // does not check for errors
+    {
+      if (c >= 'a')
+      {
+        return 10 + (c - 'a');
+      }
+
+      if (c >= 'A')
+      {
+        return 10 + (c - 'A'); 
+      }
+
+      return c - '0';
+    }
+  } // namespace
+
+  Uuid Uuid::Parse(std::string const& s)
   {
-    Uuid rv{uuid.data()};
-    return rv;
+    bool parseError = false;
+    Uuid result;
+    if (s.size() != UuidStringLength)
+    {
+      parseError = true;
+    }
+    else
+    {
+      for (size_t i = 0, j = 0; i < UuidStringLength; ++i)
+      {
+        const auto c = s[i];
+        if (IsDashIndex(i))
+        {
+          if (c != '-')
+          {
+            parseError = true;
+            break;
+          }
+        }
+        else
+        {
+          const auto c2 = s[i + 1];
```

Comment:
Add an assert that i+1 is less than UuidStringLength? Otherwise I think this might run off the end of the string (the loop invariant is on i < UuidStringLength so I believe this canaccess UuidStringLength into the string, which is 1 past the end?

Reply:
Yes, in theory it could, I see, this could be more fragile, if a developer in the future would modify this code incorrectly, I don't see how it wouldn't be caught by the tests though, but I'll see what I can do, I think I'll end up adding the AZURE_ASSERT as you're suggesting.

Reply:
I ended up adding `assert()`. I saw you are also using it. `AZURE_ASSERT()` seems like an overkill to me, because it does compile in release configurations as well, and aborts if condition is not true, which might be right in some circumstances, but I don't think it should be used here, especially that it is used inside a loop.